### PR TITLE
(#86) Extend run_example.sh to build-and-test simple_app_under_sev

### PR DESCRIPTION
**Pull request by gapisback**
This commit extends run_example.sh to build-and-test simple_app_under_sev
    sample program. However, this app requires some special handling and
    execution as 'root'. Some of that logic / checks is now coded into
    the driver script. Things work partially, and will be improved in
    subsequent commits.

SEV-app (like the OE-app) supports both manual and automated policy
    generation. This commit only implements support for the latter,
    through the 'setup_with_auto_policy_generation_for_SEV' tag.

User is expected to provide platform-specific policy-JSON file
    and appropriate certificates. The scripts expects these to be
    dropped in simple_app_under_sev/platform_data/ sub-dir. If they don't
    exist, errors will be reported, and execution aborted. These
    files are distributed by the script to the required sub-dirs,
    including the provisioning/ dir.

The measurement is grabbed from the user-provided ./sev_policy.json
    file, so it's critical that this be provided upfront.

Deleted instructions.txt and rewrote it as md-file. Reworked
    md-docs to be inline with docs for other applications.

Miscellaneous:
- Improved error messages in .cc files where some system calls fail
- Improved diagnostics from the driver script
- Improved messages from --dry-run, to see what's being run

Pending work (to be completed in future commits):
- The script recognizes this app, and runs the commands under
      'run_test' tag as 'root'. (The entire app still does not succeed
      even with this when run through the script.)

- manual_policy_generation_for_SEV() is a stub; currently not supported

### CODE REVIEW

File: sample_apps/run_example.sh [hunk 1]
```diff
@@ -47,6 +49,10 @@ Jq=jq    # Needed for OE app; will be established later on.
 MbedTLS_dir="mbedtls"
 Simple_App_under_gramine_MbedTLS_dir="sample_apps/simple_app_under_gramine/${MbedTLS_dir}"
 
+# Work-instructions for SEV-app need following files pre-populated by the user.
+#
+Platform_data_files_SEV="sev_policy.json ark_cert.der ask_cert.der vcek_cert.der"
```
Comment (gapisback) [inline sample_apps/run_example.sh:54]: NOTE: The work instructions need to be updated to mention `platform_data` and these files

File: sample_apps/run_example.sh [hunk 2]
```diff
@@ -640,6 +773,15 @@ function manual_policy_generation_for_OE() {
     run_popd
 }
 
+# ###########################################################################
+# RESOLVE: Fill this out.
+function manual_policy_generation_for_SEV() {
```
Comment (gapisback) [inline sample_apps/run_example.sh:778]: This will be fleshed out in a future change-set.

File: sample_apps/run_example.sh [hunk 3]
```diff
@@ -1142,23 +1401,42 @@ function run_simple_app_under_gramine_as_client_talk_to_Cert_Service() {
...
-    run_cmd gramine-sgx gramine_example_app \
-        --data_dir="./${Client_app_data}/"  \
-        --operation=cold-init               \
-        --policy_store_file=policy_store    \
-        --print_all=true || 0
+    run_cmd gramine-sgx gramine_example_app         \
+                --data_dir="./${Client_app_data}/"  \
+                --operation=cold-init               \
+                --policy_store_file=policy_store    \
+                --print_all=true || 0
```
Comment (gapisback) [inline sample_apps/run_example.sh:1408]: No logic changes; only indentation changes.

File: sample_apps/run_example.sh [hunk 4]
```diff
@@ -1142,23 +1401,42 @@ function run_simple_app_under_gramine_as_client_talk_to_Cert_Service() {
...
-    run_cmd gramine-sgx gramine_example_app \
-        --data_dir="./${Client_app_data}/"  \
-        --operation=get-certifier           \
-        --policy_store_file=policy_store    \
-        --print_all=true || 0
+    run_cmd gramine-sgx gramine_example_app         \
+                --data_dir="./${Client_app_data}/"  \
+                --operation=get-certifier           \
+                --policy_store_file=policy_store    \
+                --print_all=true || 0
```
Comment (gapisback) [inline sample_apps/run_example.sh:1416]: No logic changes; only indentation changes.

File: sample_apps/run_example.sh [hunk 5]
```diff
@@ -1290,11 +1587,19 @@ function run_simple_app_under_gramine_as_client_make_trusted_request() {
     echo "${Me}: $(TZ="America/Los_Angeles" date) Completed ${FUNCNAME[0]} ${DryRun_msg}."
 }
 
+# ###########################################################################
+function run_simple_app_under_sev_as_client_make_trusted_request() {
```
Comment (gapisback) [inline sample_apps/run_example.sh:1591]: Relocate to L1534.

**Comment by gapisback** [comment]
@yelvmw -- Although the entire script is not running cleanly e2e for the interfaces implemented, I think this code is ready for review.

As it's a biggish change-set, I'd like to clean up base on review feedback and integrate this to `/main`. We may end up having to simply document that they have to run some steps one at a time, and `run_test` target will not succeed on its own (due to `sudo` issues).

**Comment by gapisback** [comment]
NOTE: I am going thru the md-file myself, and already found some typos. Will clean this up while the review is going on, so you can see delta-diffs from edits I've made.

I think the instructions are very complex ... with this manual / auto and simulated modes.
We should talk on Tuesday to straighten these out.

**Review by jlmucb — APPROVED** [review]

**Review by jlmucb — APPROVED** [review]

File: sample_apps/run_example.sh (hunk 1 above)
Comment (gapisback) [reply sample_apps/run_example.sh:54]: Done. Added a note to instructions.md about these 4 files.

File: sample_apps/run_example.sh (hunk 5 above)
Comment (gapisback) [reply sample_apps/run_example.sh:1594]: Done. relocated call to just before callee, so it's clear that both `simple_app` and `simple_app_under_sev` run the same shared function.

**Comment by gapisback** [comment]
Final re-test after all rebases is also passing: (Attn: @yelvmw )

```
run_example.sh: Running run_app_by_name_as_client_make_trusted_request
+ set +x

+ sudo /home/rgerganov/agurajada/certifier-framework-for-confidential-computing/sample_apps/simple_app_under_sev/sev_example_app.exe --data_dir=./app1_data/ --operation=run-app-as-client --policy_store_file=policy_store --print_all=true
running as client
Server peer id is Measured-f9cbbae12fe97fbbb325498d5c79eddec0aea9e8cf47087cb0bacd2f29c5cb80d03eef4dcfabcde7737e705d4d1b2164
Client peer id is Measured-f9cbbae12fe97fbbb325498d5c79eddec0aea9e8cf47087cb0bacd2f29c5cb80d03eef4dcfabcde7737e705d4d1b2164
SSL server read: Hi from your secret client

SSL client read: Hi from your secret server

+ set +x
```

File: sample_apps/simple_app_under_sev/sev_policy.json [hunk 6]
```diff
@@ -19,7 +19,7 @@
           "comparator": "eq",
           "type": "string",
           "name": "smt",
-          "value": "no"
+          "value": "yes"
```
Comment (yelvmw) [inline sample_apps/simple_app_under_sev/sev_policy.json:22]: Let's not commit these changes. The default ones here are for the simulator and should be preserved.

File: src/sev-snp/sev_support.cc [hunk 7]
```diff
@@ -46,6 +46,7 @@
 using namespace certifier::framework;
 using namespace certifier::utilities;
 
+// Had to sudo chmod 644 this to allow non-root to open this device
```
Comment (yelvmw) [inline src/sev-snp/sev_support.cc:49]: Let's remove this comment. It's not necessary.

File: sample_apps/simple_app_under_sev/sev_policy.json (hunk 6 above)
Comment (gapisback) [reply sample_apps/simple_app_under_sev/sev_policy.json:22]: Done. I've withdrawn the entire file and all the changes. Will re-test.

File: src/sev-snp/sev_support.cc (hunk 7 above)
Comment (gapisback) [reply src/sev-snp/sev_support.cc:49]: Done. (Oversight ...)

File: sample_apps/simple_app_under_sev/sev_policy.json (hunk 6 above)
Comment (gapisback) [reply sample_apps/simple_app_under_sev/sev_policy.json:22]: Re-tested. Script still runs cleanly.

**Review by yelvmw — APPROVED** [review]
LGTM

**Pull request merged**